oscore-cf/src/main/java/org/eclipse/californium/oscore/Decryptor.java

```diff
@@ -0,0 +1,272 @@
+	 * @param seq the sequence number
+	 * 
+	 * @return the serialized AAD
+	 */
+	protected static byte[] serializeAAD(Message message, OSCoreCtx ctx, int seq) {
+		if (message instanceof Request) {
+			Request r = (Request) message;
+			return OSSerializer.serializeReceiveRequestAAD(CoAP.VERSION, seq, ctx, r.getOptions());
+		} else if (message instanceof Response) {
+			Response r = (Response) message;
+			return OSSerializer.serializeReceiveResponseAAD(CoAP.VERSION, seq, ctx, r.getOptions());
+		}
+		return null;
+	}
+
+	/**
+	 * Decompress the message.
+	 * 
+	 * @param cipherText the encrypted data
+	 * @param message the received message
+	 * @return the Encrypt0Message
+	 * @throws OSException
+	 */
+	protected static Encrypt0Message decompression(byte[] cipherText, Message message) throws OSException {
+		Encrypt0Message enc = new Encrypt0Message(false, true);
+
+		decodeObjectSecurity(message, enc);
+
+		if (cipherText != null)
+			enc.setEncryptedContent(cipherText);
+		return enc;
+	}
+
+	/**
+	 * Decodes the Object-Security value.
+	 * 
+	 * @param message the received message
+	 * @param enc the Encrypt0Message object
+	 * @throws OSException
+	 */
+	private static void decodeObjectSecurity(Message message, Encrypt0Message enc) throws OSException {
+		byte[] total = message.getOptions().getOscore();
+
+		byte flagByte = total[0];
+
+		int n = flagByte & 0x07;
+		int k = flagByte & 0x08;
+		int h = flagByte & 0xa6;
+
+		byte[] partialIV = null;
+		byte[] kid = null;
+		int index = 1;
+
+		if (n > 0) {
+			try {
+				partialIV = Arrays.copyOfRange(total, index, index + n);
+				index += n;
+			} catch (Exception e) {
+				LOGGER.error("Partial_IV is missing from message when it is expected.");
+				throw new OSException(Error.FAILED_TO_DECODE_COSE);
+			}
+		}
+
+		if (h == 16) {
+			int s = total[index];
+			index += 1;
+
+			if (s > 0) {
+				LOGGER.error("Kidcontext is included, but it is not supported. We ignore it and continue processing.");
+			} else {
+				LOGGER.error("Kid context is missing from message when it is expected.");
+				throw new OSException(Error.FAILED_TO_DECODE_COSE);
+			}
+		}
+
+		if (k == 8) {
+			kid = Arrays.copyOfRange(total, index, total.length);
+		} else {
+			if (message instanceof Request) {
+				LOGGER.error("Kid is missing from message when it is expected.");
+				throw new OSException(Error.FAILED_TO_DECODE_COSE);
+			}
+		}
+
+		try {
+			if (partialIV != null)
+				enc.addAttribute(HeaderKeys.PARTIAL_IV, CBORObject.FromObject(partialIV), Attribute.UNPROTECTED);
+			if (kid != null)
+				enc.addAttribute(HeaderKeys.KID, CBORObject.FromObject(kid), Attribute.UNPROTECTED);
+		} catch (CoseException e) {
+			e.printStackTrace();
+		}
+	}
+
+	/**
+	 * Replaces the message's options with a new OptionSet which doesn't contain
+	 * any of the non-special E options as outer options
+	 * 
+	 * @param message
+	 */
+	protected static void discardEOptions(Message message) {
```

Comment:
Should E options really be silently removed for received messages?

Reply:
Yes. E options which are put as outer options and which are not supposed to be outer options are removed early on when receiving response/request.

Reply:
Adding logging for the action.

Reply:
I still miss the logging.

Reply:
I had placed the logging in the decryptor. Will fix.